### PR TITLE
Brand new duplicate feature(s) and move feature(s) functionality

### PR DESCRIPTION
**Pull request by nirvn**
This PR adds a feature duplication and feature geometry move functionality into QField.

The new functionalities can be access via the feature list's multi feature selection mode menu as well as the feature form's menu. Here's a screencast showcasing the new functionalities:

https://user-images.githubusercontent.com/1728657/135201588-f836d985-9202-4d82-bf31-32edcc25f245.mp4

Implementation note: when features are duplicated, QField resets the feature list to show the duplicated feature(s), pre-selected. This insures that follow up actions (such as moving features or multi editing features) is done against the duplicated features and not the original features.

Icons were added after the above-uploaded video was recorded. Here's how they look like:
image

### CODE REVIEW

**Comment by nirvn** [comment]
BTW, this is nice to see our migration towards a menu for these areas on QField is paying of. We can quite easily expand functionality without cluttering the UI. Win-win.

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for f0d704372a7e43906f0aa36551f76768e1a2db50:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-22853c5c50021f823293915a4166d31ea97c8c18-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-22853c5c50021f823293915a4166d31ea97c8c18-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-22853c5c50021f823293915a4166d31ea97c8c18-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-22853c5c50021f823293915a4166d31ea97c8c18-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 16b880f6027e95278bec43c96476a3b3fc486692:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-c47818f697666962825b0ef41d06a05b03a4247c-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-c47818f697666962825b0ef41d06a05b03a4247c-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-c47818f697666962825b0ef41d06a05b03a4247c-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-c47818f697666962825b0ef41d06a05b03a4247c-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 2254557fb02d3f3e9a3a0f014381722b83d62a63:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-ff11e5933cc2fbd38a9dad8fc3599f0624e50420-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-ff11e5933cc2fbd38a9dad8fc3599f0624e50420-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-ff11e5933cc2fbd38a9dad8fc3599f0624e50420-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-ff11e5933cc2fbd38a9dad8fc3599f0624e50420-x86.apk)

File: src/core/utils/layerutils.cpp [hunk 1]
```diff
@@ -203,3 +204,45 @@ bool LayerUtils::deleteFeature( QgsProject *project, QgsVectorLayer *layer, cons
...
+    return QgsFeature();
+  }
+
+  QgsFeature duplicatedFeature;
+  QObject tmp;
+  auto featureAdded = [ layer, &duplicatedFeature ]( QgsFeatureId fid )
+  {
+    duplicatedFeature = layer->getFeature( fid );
+  };
+  connect( layer, &QgsVectorLayer::featureAdded, &tmp, featureAdded );
```
Comment (m-kuhn) [inline src/core/utils/layerutils.cpp:228]:
```suggestion
  auto featureAdded = [ layer, &duplicatedFeature ]( QgsFeatureId fid )
  {
    duplicatedFeature = layer->getFeature( fid );
  };
  QMetaObject::Connection connection = connect( layer, &QgsVectorLayer::featureAdded, featureAdded );
  auto sweaper = qScopeGuard([layer, connection] { layer->disconnect( connection ); });
```

just because

File: src/core/utils/qfieldcloudutils.cpp [hunk 2]
```diff
@@ -65,7 +65,7 @@ const QString QFieldCloudUtils::getProjectId( const QString &fileName )
   QString basePath = QFileInfo( baseDir.path() ).canonicalFilePath();
   QString cloudPath = QFileInfo( localCloudDirectory() ).canonicalFilePath();
 
-  if ( basePath.startsWith( cloudPath ) )
+  if ( !cloudPath.isEmpty() && basePath.startsWith( cloudPath ) )
```
Comment (m-kuhn) [inline src/core/utils/qfieldcloudutils.cpp:68]: CC @suricactus does that look safe (seems unrelated to this feature)

**Comment by m-kuhn** [comment]
Nice work, good for me after mostly a feedback from @suricactus and maybe my code-simplixification

File: src/core/utils/qfieldcloudutils.cpp (hunk 2 above)
Comment (suricactus) [reply src/core/utils/qfieldcloudutils.cpp:68]: Yes, it is. Cannonical path retuns empty string on the first clear install on a new device, which causes the `if` to fail (`"".startsWith("path/here")`). However, if you download your first project, all works fine. That's why it was hard to catch it and it was there for ages. Thanks to @nirvn that he caught it and for the fix.

File: src/core/utils/layerutils.cpp (hunk 1 above)
Comment (nirvn) [reply src/core/utils/layerutils.cpp:228]: @m-kuhn , not just because, that's what I'd call doing the right thing :)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for cc70a497c5eee1efd831eb0c142151acaec5fb4a:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-93c4fa5706374cffd509f7e8de8b3620703de364-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-93c4fa5706374cffd509f7e8de8b3620703de364-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-93c4fa5706374cffd509f7e8de8b3620703de364-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-93c4fa5706374cffd509f7e8de8b3620703de364-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for da2561f609e085de8d2da3a1ff5066b821a08844:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-ac1a1e6f17c6e45e6c20d1f040732bf871b0c646-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-ac1a1e6f17c6e45e6c20d1f040732bf871b0c646-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-ac1a1e6f17c6e45e6c20d1f040732bf871b0c646-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-ac1a1e6f17c6e45e6c20d1f040732bf871b0c646-x86.apk)

**Comment by mbernasocchi** [comment]
Task linked: [DEV-3115 Duplicate and move points](https://app.clickup.com/t/2192114/DEV-3115)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for da2561f609e085de8d2da3a1ff5066b821a08844:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-32f59e20aabe7e631ae64a90468bef254ba51376-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-32f59e20aabe7e631ae64a90468bef254ba51376-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-32f59e20aabe7e631ae64a90468bef254ba51376-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-32f59e20aabe7e631ae64a90468bef254ba51376-x86.apk)

File: src/core/utils/layerutils.cpp [hunk 3]
```diff
@@ -203,3 +206,44 @@ bool LayerUtils::deleteFeature( QgsProject *project, QgsVectorLayer *layer, cons
...
+    QgsMessageLog::logMessage( tr( "Cannot start editing" ), "QField", Qgis::Warning );
+    return QgsFeature();
+  }
+
+  QgsFeature duplicatedFeature;
+  QMetaObject::Connection connection = connect( layer, &QgsVectorLayer::featureAdded, [ layer, &duplicatedFeature ]( QgsFeatureId fid )
+  {
+    duplicatedFeature = layer->getFeature( fid );
+  } );
+  auto sweaper = qScopeGuard( [ layer, connection ] { layer->disconnect( connection ); } );
```
Comment (suricactus) [inline src/core/utils/layerutils.cpp:229]: Nice one!

File: src/core/utils/layerutils.cpp [hunk 4]
```diff
@@ -203,3 +206,44 @@ bool LayerUtils::deleteFeature( QgsProject *project, QgsVectorLayer *layer, cons
...
+    QgsMessageLog::logMessage( tr( "Cannot start editing, no layer" ), "QField", Qgis::Warning );
+    return QgsFeature();
+  }
+
+  if ( !layer->startEditing() || !layer->editBuffer() )
+  {
+    QgsMessageLog::logMessage( tr( "Cannot start editing" ), "QField", Qgis::Warning );
+    return QgsFeature();
+  }
+
```
Comment (suricactus) [inline src/core/utils/layerutils.cpp:223]:
```suggestion
  if ( !feature.isValid() )
  {
    QgsMessageLog::logMessage( tr( "Cannot copy invalid feature" ), "QField", Qgis::Warning );
    return QgsFeature();
  }
```
Does it make sense to check the feature validity too?

File: src/core/utils/layerutils.h [hunk 5]
```diff
@@ -60,6 +60,8 @@ class LayerUtils : public QObject
 
     static bool deleteFeature( QgsProject *project, QgsVectorLayer *layer, const QgsFeatureId fid, bool shouldWriteChanges = true );
 
+    static QgsFeature duplicateFeature( QgsVectorLayer *layer, const QgsFeature &feature );
```
Comment (suricactus) [inline src/core/utils/layerutils.h:63]: Some docs, please.

**Review by suricactus — COMMENTED** [review]
Looks really cool, thanks @nirvn !

Played with in on my phone, also looks solid. I got one single crash in the begining, but did not manage to reproduce.

**Comment by suricactus** [comment]
Screenshot_2021-10-01-22-15-52-13_13baf0361b171d856e788683d2e49628

**Comment by suricactus** [comment]
https://user-images.githubusercontent.com/2820439/135679419-b6de967d-35ad-4172-957d-ad48bb5148c6.mp4

File: src/core/multifeaturelistmodelbase.cpp [hunk 6]
```diff
@@ -320,6 +319,15 @@ bool MultiFeatureListModelBase::canDeleteSelection()
   return !vlayer->readOnly() && ( vlayer->dataProvider()->capabilities() & QgsVectorDataProvider::DeleteFeatures ) && !vlayer->customProperty( QStringLiteral( "QFieldSync/is_geometry_locked" ), false ).toBool();
 }
 
+bool MultiFeatureListModelBase::canDuplicateSelection()
+{
+  if ( mSelectedFeatures.isEmpty() )
+    return false;
+
+  QgsVectorLayer *vlayer = mSelectedFeatures[0].first;
```
Comment (suricactus) [inline src/core/multifeaturelistmodelbase.cpp:327]: Isn't it quite optimistic to think all the selected features are the same layer? Probably needs docs and ASSERT?

**Review by suricactus — CHANGES_REQUESTED** [review]
I think we need to fix these crashes before merge.

File: src/core/utils/layerutils.cpp (hunk 3 above)
Comment (nirvn) [reply src/core/utils/layerutils.cpp:229]: Credit to @m-kuhn

File: src/core/multifeaturelistmodelbase.cpp (hunk 6 above)
Comment (nirvn) [reply src/core/multifeaturelistmodelbase.cpp:327]: Added a note, and fixed the issue above (whereas a duplicated feature would lead to an out-of-sync model filter / selected features' layer).

**Comment by nirvn** [comment]
@suricactus , to avoid the potential pain of conflicts, I've address your review in PR #2155.

**Comment by nirvn** [comment]
Alright, git had another plan, both features merged into this PR.

@suricactus , the crash you identified is fixed, so is the out-of-sync layer filtering when features are selected. Thanks for the reviews.

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 2210ac503467435019c6fa836972f2bb9648c904:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-da44a7b5a1d8dc61baa0679bb367b48613e91066-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-da44a7b5a1d8dc61baa0679bb367b48613e91066-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-da44a7b5a1d8dc61baa0679bb367b48613e91066-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-da44a7b5a1d8dc61baa0679bb367b48613e91066-x86.apk)

**Comment by nirvn** [comment]
@suricactus , PR rebased against the fixes approved just now, inc. the crasher referred to above.

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 5f7f286c1c5a128ffe9bb1ad69b72743e1e41a61:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-ec49f4f9a70507f5be294f692829f85f3d1dd4e0-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-ec49f4f9a70507f5be294f692829f85f3d1dd4e0-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-ec49f4f9a70507f5be294f692829f85f3d1dd4e0-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-ec49f4f9a70507f5be294f692829f85f3d1dd4e0-x86.apk)

**Comment by suricactus** [comment]
https://user-images.githubusercontent.com/2820439/135712479-ed1817b5-7490-40ea-869a-e73e74e2811d.mp4


There might be one more, let's fix this first. I got a crash after duplicating too, but not sure if related to this.

**Comment by nirvn** [comment]
@suricactus , crash fixed.

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for b6bc5b32c72935b4b5efa2911466bd9328166ce6:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-bef8152d080ddcfb62409289308a118cf2249b58-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-bef8152d080ddcfb62409289308a118cf2249b58-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-bef8152d080ddcfb62409289308a118cf2249b58-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2146-bef8152d080ddcfb62409289308a118cf2249b58-x86.apk)

**Comment by suricactus** [comment]
Just tested. Awesome work @nirvn , can't wait to see this on master.

**Review by suricactus — APPROVED** [review]
